Added ability to modify IOU Amount

sig5 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
The NumberPad was missing handling for some features, Added handling for different use cases. We are modifying the selection prop via the selection API  and optimistic updates are being done in `IOUAmount.js`
Note:
Due to the development overheads, the dev App might have some cursor inconsistencies in the dev build. The App works fine in the release builds.
Edit: The components have been restructured such that:
1. iOS uses controlled selection.
2. Rest of the platforms use uncontrolled selection.

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/6154
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose and it is
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [x] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.


<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### Tests
- [X] Verify that no errors appear in the JS console

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Go to Send/Receive Money for a chat.
2. Test the following use cases

- Delete a block of text after selecting.
- Delete a single character after entering.
- Select a set of characters and replace it with another character.
- Select one character and replace it with any other character.
- Check if the input allows for invalid values.
- Try copying and pasting valid text in the input.

3. Verify that correct values appear for the operations.
- [X] Verify that no errors appear in the JS console

### Tested On

- [X] Web
- [X] Mobile Web
- [x] Desktop
- [x] iOS
- [X] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web 
<!-- Insert screenshots of your changes on the web platform-->

https://user-images.githubusercontent.com/48352577/151290358-276fd900-573f-409f-bca3-c37635783430.mov


#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
https://user-images.githubusercontent.com/48352577/151291211-afb3d566-272d-469b-bd2a-9e2a1f343638.mp4


#### Desktop 
<!-- Insert screenshots of your changes on the desktop platform-->

https://user-images.githubusercontent.com/48352577/152880990-38313f03-3566-491a-a476-c4f0bc4442e2.mp4

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

https://user-images.githubusercontent.com/48352577/152880887-6921e413-a9d9-44dd-9bab-6d83369ea37e.mp4




#### Android
<!-- Insert screenshots of your changes on the Android platform-->

https://user-images.githubusercontent.com/48352577/151290399-3da37672-501d-4dfe-b565-c37dc6bb73ac.mp4

sig5 committed:
Added textInput selection and amount handling

sig5 committed:
added newline at the end

sig5 committed:
updated comment

sig5 commented:
@rushatgabhane Could you please test it on iOS, I don't have it available atm.

sig5 committed:
refactor vars

sig5 committed:
adjusted spacing

sig5 committed:
removed stray comments

sig5 commented:
@rushatgabhane I do not have access to an iOS device rn, So I couldn't test it there.I hope you can verify if it works :D

rushatgabhane commented:
@sig5 The cursor position isn't being updated on iOS, can you find a fix for that?

sig5 commented:
Hi folks, sorry for the delay. I have been trying to find out the reason for this erratic behavior on iOS. I will list down the approaches and issues with them down below:
1. The initial solution which involves the standard approach of keeping `selection` as a state variable and updating state on `onSelectionChange` Callback seems to work fine on both platforms. But there is a **BUG** on android. The issue thread mentions it. The copy-paste menu tends to get hidden on Android devices. This solution seems to work fine on iOS though.

2.  The approach we took to counter the aforementioned bug is to set the `selection` on the TextInput using `setNativeProps` (i.e the selection API).. The approach seemed to work fine on the Android devices, but the iOS devices perform poorly. The selection value of the native element doesn't seem to update as expected. The issue though seems to be internal to  `react-native` as the `setNativeProps` calls seem to not work only on ios devices for us. I have tested this on a dev build. We have observed that the cursor hopping bug(an unexpected behavior) was due to an over bundled build and did not appear on the Release variant. I do not have access to an Apple developer account, so I can't test this PR on the `Release` build, but it might be worthwhile to test if the results are inconsistent on a `release` variant too.

I am trying to add find any more suitable solution if I can. Currently, the two available approaches work exclusively on iOS and Android respectively.

rushatgabhane commented:
> I am trying to add find any more suitable solution if I can

@sig5 let me know if you have any updates, thanks.

sig5 committed:
seperated out ios and native

sig5 commented:
Hi guys,
It seems like the iOS selection API might have some issues here on iOS. The `setNativeProps` does not play as expected. These issues might also affect our future plans to make `TextInputs` uncontrolled for better performance. Below are the observations(The behavior as expected in Android devices).
1) The selection updates work as expected if the `text` prop of the Input element stays unchanged. Relevant snack : [link](https://snack.expo.dev/@sig5/spunky-pizza)

2) The first unexpected behavior occurs on the `TextInput` when we start updating the text as well as selection and s`election.start ===selection.end.` The selection prop of `TextInput` doesn't get updated at all. Snack: [link](https://snack.expo.dev/@sig5/add0b7)

3) When testing I found that the index placements of the API can be unexpected too. The immovable cursor problem can be resolved up to an extent by setting the index at` { start: expectedStart-1, end: expectedEnd-1}`(which is unexpected and found while trying to debug the code) . But this feels like a workaround.

The issue @rushatgabhane mentioned in the previous comment was faced here too: https://github.com/Expensify/App/issues/6876#issuecomment-1000496096.

rushatgabhane commented:
@sig5 sorry, will get back to you tomorrow.

rushatgabhane commented:
@sig5 thanks for your effort, and attaching some snack links. Although, a video would be preferred next time.

> The unexpected behavior occurs on the TextInput when we start updating the text prop as well as selection prop and selection.start !==selection.end. The selection prop of TextInput doesn't get updated to the expected value. Snack: [link](https://snack.expo.dev/@sig5/add0b7)

Can you help me with a case where this is applicable to the issue or anytime in future? 

> The selection updates work as expected if the text prop of the Input element stays unchanged.

Do you think direct manipulation of value using setNativeProps will help?

sig5 commented:
Hi @rushatgabhane ,
>@sig5 thanks for your effort, and attaching some snack links. Although, a video would be preferred next time.

I added snack links because this seems like a react native issue and wanted to demonstrate the same. :D will take care of it from now on.
>Can you help me with a case where this is applicable to the issue or anytime in the future?

This issue doesn't concern us as of now, but I have mentioned it if we come across it in the future. The behavior of the selection prop is unexpected in the other cases too.
> Do you think direct manipulation of value using setNativeProps will help?

I have tried to do that. But that doesn't help.

rushatgabhane commented:
Okay, I see what you mean.
iOS selection is broken when we set selection using native native method. But it works fine when we pass it as a prop.

> We should possibly segregate the logic for iOS using the passing down the props approach and open a tracking issue in the react-native library.

I'd like us to set selection using native props as it's faster but we can't do that for iOS yet. 
So let's create a seperate iOS file which uses the selection prop to fix this issue. I hope that makes sense

Could you also link/create a relevant react-native issue?

sig5 commented:
> So let's create a seperate iOS file which uses the selection prop to fix this issue. I hope that makes sense

Yes, that will be ideal.

> Could you also link/create a relevant react-native issue?

Sure I will do that.

sig5 committed:
seperated ios file

sig5 committed:
moved to a folder

sig5 committed:
aggregated proptypes

sig5 committed:
Seperated common utility functions

sig5 committed:
Seperated common utility functions

sig5 committed:
fixed binding issues

sig5 committed:
Added base page

sig5 committed:
Added base page

sig5 committed:
Added base page

sig5 committed:
modified ios for base page props

sig5 committed:
Removed utils file

sig5 commented:
Requested review by mistake, will do a couple of changes and re-request reviews.

sig5 committed:
fixed ios issue

rushatgabhane commented:
@sig5 just a reminder to merge `Expensify:main` into your branch because there are some conflicts.

sig5 committed:
restructured components

sig5 committed:
Moved proptypes

sig5 committed:
renamed component

sig5 committed:
reverted base rename

sig5 committed:
fixed imports

sig5 committed:
Added proptypes

sig5 committed:
resolved merge conflict

sig5 committed:
resolved conflict

sig5 commented:
Updated PR

sig5 committed:
fixed linting issue

rushatgabhane reviewed (changes requested):
@sig5 you've missed this edge case - amount can be more than 9 digits.


https://user-images.githubusercontent.com/29673073/153026737-841aa678-d5d6-4645-8a5b-e3b270160c33.mp4

rushatgabhane reviewed (changes requested):
Hey @sig5, I could also find this bug.
On android, 
1. Place the cursor in middle of some amount. 
2. Hold backspace.

Numbers after the cursor also get erased.


https://user-images.githubusercontent.com/29673073/153028781-3f5ed1e9-d998-46b5-bef3-8b3b08a93843.mp4

sig5 commented:
> Hey @sig5, I could also find this bug. On android,
> 
> 1. Place the cursor in middle of some amount.
> 2. Hold backspace.
> 
> Numbers after the cursor also get erased.
> 
>  screen-20220208-190825.mp4

Hi Rushat, can you test it on a release build? It seems like overbundling encountered only on dev version. I can't reproduce it on release variant.

sig5 committed:
Handled edge case

sig5 commented:
> @sig5 you've missed this edge case - amount can be more than 9 digits.
> 
>  screen-20220208-190253.mp4

Handled!

rushatgabhane reviewed (changes requested):
Nice work, requesting really small style changes.

rushatgabhane reviewed (changes requested):
Hey @sig5, sorry but selectionEnd isn't visible for mWeb.
Could you find a fix for this?

Steps
1. Type 8 digit amount
2. Tap and hold to select the amount
<img src="https://user-images.githubusercontent.com/29673073/153041558-1a61de17-91bc-465e-a00e-335acea84051.png" height=600>

sig5 commented:
> Hey @sig5, sorry but selectionEnd isn't visible for mWeb. Could you find a fix for this?
> 
> Steps
> 
> 1. Type 8 digit amount
> 2. Tap and hold to select the amount
> 
> <img alt="" height="600" src="https://user-images.githubusercontent.com/29673073/153041558-1a61de17-91bc-465e-a00e-335acea84051.png">

@rushatgabhane I can't reproduce it on my device, I am using Chrome Mobile too. Can it be device-specific?

sig5 committed:
fixed formatting

sig5 commented:
I updated and I still can't reproduce. What's interesting is I can't even reproduce it on a Pixel 3 Emulator
![Screenshot from 2022-02-09 02-23-16](https://user-images.githubusercontent.com/48352577/153073799-8af4303b-15d9-4fc4-b049-c84c34b8499b.png)
.

sig5 committed:
Added newline

rushatgabhane commented:
@sig5 thanks! Looks like it's a Chrome issue only on the latest version, because it isn't repro on previous versions. 
And Chrome 99 beta doesn't have this problem.

rushatgabhane reviewed (changes requested):
@sig5 iOS is using a different implementation. Let's add a detailed comment on top of iOS file explaining why so.

Also, let's link the react native issue in the file too. Let me know what you think, thanks so much!

sig5 commented:
Hi @rushatgabhane , That will be ideal. Dropping the link to issue here too: [Issue](https://github.com/facebook/react-native/issues/33077)

sig5 committed:
Added comment

sig5 committed:
updated comment

rushatgabhane commented:
@sig5 this has been a long one mainly due to the react native bug for iOS, I appreciate for you patience.
I'll go over this once again, and test all platforms by the end of the day. 

Cheers!

rushatgabhane reviewed (dismissed):
@puneetlath LGTM! Tests well on all platforms!
 
But let's merge https://github.com/Expensify/App/pull/7318 first as it's a refactor, and let this PR handle the conflicts.

puneetlath commented:
@sig5 please merge main and then test your PR again. I'll be happy to give it a final review then!

sig5 committed:
Merge branch 'main' into fix-#6154

sig5 committed:
Removed IOUAmountPage.js

sig5 committed:
Added merge changes

sig5 commented:
Hi @puneetlath unfortunately the PR [7318](https://github.com/Expensify/App/pull/7318) breaks our updated code. The BaseTextInput updates value via `setNativeProps` which reproduces our iOS issue. 
@parasharrajat  Is making selection uncontrolled the most necessary thing to do in #7318? It leads us to reproduce the [Issue](https://github.com/facebook/react-native/issues/33077). This will most probably lead to inconsistent results on iOS when selection is manually handled.

cc- @rushatgabhane

parasharrajat commented:
Sorry, I am clueless here. Can someone point me to the right comments to understand the issue and the patch applied to fix it?

sig5 commented:
Hi @parasharrajat , We happen to encounter caret issues on iOS when selection is manually handled and `setNativeProps` is used (either for `selection` or `value`). So we decided to fall back to use the state to pass down the text and selection on `iOS` platform.
Concisely:
 On iOS only -  The following cases lead to cursor being stuck.
* setting selection using `setNativeProps`, and value using `setNativeProps`.
* setting selection using props, and value using `setNativeProps`.

Now since `basetextInput` is setting value using `setNativeProps` in `ComponentdidUpdate`, We again wind up with the same situation. 

- Can we do away with using `setNativeProps`? OR
- Can we introduce a new prop that allows us to choose if we want to use `setNativeProps` or we would like to pass value down as a prop.

parasharrajat commented:
Thanks, @sig5. Appreciate the summary. I am happy to investigate this issue but I need another help from you. 
Could you please create a copy of this branch and resolve the conflicts there and give me the branch URL on GH? I will pull down the branch and try to debug it. 

Note: I am really busy these days so It could take me some time before I can give you my conclusion. Thankfully weekend(~winter~) is coming, so I would be able to spend time on it. ETA - Tuesday.

sig5 commented:
Thanks, @parasharrajat . I added changes in my fork here: https://github.com/sig5/App/tree/testing-6154

IOUAmountPage is now split for iOS and the rest of the platforms. Good luck with this winter xD.

parasharrajat commented:
I have some results but they could be disheartening. It could be the problem with react-native IOS implementation. But mostly playing with selection is like putting hands in the bone fire. 

I have observed a few strange things and some improvements. First, let's run some tests. 
Note: I removed index.ios.js file for IOUAMOUNT.js to run these tests. And I could have missed some steps in the below test as I was not taking notes during testing.

1. I removed the setnativeProps call from https://github.com/Expensify/App/blob/8431ed41e33092af1e42a143d7364f95b14287fe/src/components/TextInput/BaseTextInput.js#L66 line. And still text input was updating when we remove the entered values. How? I want to know why? we are not using `value` prop instead `defaultValue`. If you find out let me know.  It could be an indication that we are unnecessary re-rendering the TextInput.

2. Now, when I removed this line, the cursor jumping issue was still present on IOS. Something tells me that it does not have to do anything with `setNativeProps` theory. 

3. Then I removed the https://github.com/sig5/App/blob/f50a882713fec3c5aef4d76fbcbfdf7f95695bb7/src/pages/iou/steps/IOUAmountPage/IOUAmount/index.js#L72. And Most of the selection setting was working naturally. Only when the user selects a part of the value and clicks `<` it does not work.

4. From point 3, we notice that there is a natural cursor placement strategy from the platform which **could** be interfering with our selection setting. 

### Oberservations. 

1. This line is interesting. https://github.com/sig5/App/blob/f50a882713fec3c5aef4d76fbcbfdf7f95695bb7/src/pages/iou/steps/IOUAmountPage/index.js#L90. 

Here we are handling the Backspace as well which is automatically applied from the platform as well. But we are not preventing native events. (Why?)
 
 2. https://github.com/sig5/App/blob/f50a882713fec3c5aef4d76fbcbfdf7f95695bb7/src/pages/iou/steps/IOUAmountPage/index.js#L91 We don't have to update the selection here. It should be naturally updated. So we can skip setting selection for this case. 
 
 3. https://github.com/sig5/App/blob/f50a882713fec3c5aef4d76fbcbfdf7f95695bb7/src/pages/iou/steps/IOUAmountPage/index.js#L98 If we are removing values then why do we need to validate the amount.  And https://github.com/sig5/App/blob/f50a882713fec3c5aef4d76fbcbfdf7f95695bb7/src/pages/iou/steps/IOUAmountPage/index.js#L106
 
 4. https://github.com/sig5/App/blob/f50a882713fec3c5aef4d76fbcbfdf7f95695bb7/src/pages/iou/steps/IOUAmountPage/IOUAmount/index.js#L67-L75. 
 
 Here if we pay attention to react lifecycle, we will find that selection is set before the amount is actually updated on the input. Can it be causing inconsistencies? How? the amount is updated via state and selection is set to value `setNativeProps`. In reality, selection will be set first, and then, as state update is async, will take place. **This could be causing the bug.**
 
 
 So I tried moving the logic from [line](https://github.com/sig5/App/blob/f50a882713fec3c5aef4d76fbcbfdf7f95695bb7/src/pages/iou/steps/IOUAmountPage/IOUAmount/index.js#L72) to setState complete callback. But still no luck. 
 
 Then I wrapped https://github.com/sig5/App/blob/f50a882713fec3c5aef4d76fbcbfdf7f95695bb7/src/pages/iou/steps/IOUAmountPage/IOUAmount/index.js#L72 in **SetTimeout and everything started to work properly.**. 
 
 If we remove setting the selection ourselves. You will notice the platform automatically move the cursor to the end of the input value. (or somewhere). So either we should use the Controlled component to set both the selection and value or fully uncontrolled. There is a small [note](https://reactnative.dev/docs/direct-manipulation#avoiding-conflicts-with-the-render-function) in docs that are telling the same story as this bug but they(RN folks) didn't have time to provide an example (I guess).

sig5 commented:
Hi @parasharrajat  these are some interesting observations.
Let me answer some of your concerns.
1. The reason for this is we are still passing `inputProps` which contain value prop from `IOUAmount` . We might want to make a point to omit value prop from baseTextInput if you necessarily want to omit value prop from being passed down.

2. The interesting thing is cursor jumping issue is generally not reproducible in our release builds. Also since observably, the results of iOS implementation were really different from that of Android, we came down to the conclusion about iOS setNativeProps being buggy. If we go about tackling the problems in an iterative manner, then we should focus on the immovable cursor problem first IMHO  Also if we go back to the PR state when #7318 was not merged, it seems like we were able to handle the iOS platform in an expected manner. I feel like using `setNativeProps`  can be a problem for us.

3. This is expected. The problem is maintaining selection seems to be necessary atm for updates. Natively handled selection will work fine but when update operations are performed from the `IOUAmount` component, I reckon we will definitely run into problems.

4. I agree, but we should also agree on the fact that the results should be operationally consistent across platforms. So that calls for a bug in React -Native.

**About Observations**

1. Well, the thing with using a custom keyboard is that we aren't at all making use of native updates.

- What we are doing is: Calculate the new value as per the key > Update in on the `textInput`.
- The Custom Keyboard operations do not concern any native events.

2. and 3. I agree and these certainly are good optimizations. I will modify it in subsequent oommits

4. I have tried having the state and amount updated at the same time, but no luck.

I was occupied for a couple of days, I will try modifying the BaseTextInput to our needs in iOS. Let's see if that works.
Let me know if you  have some insights about these points :D

sig5 commented:
@parasharrajat 
> It means that we missed one case in this PR. On the Web, we leave the TextInput to native events. So I think we need to handle that as well. Please do test the cursor thing there as well.

If I am not wrong, we should not restrict native events for web. The issue is, we can have devices where we would like both 
Numeric Keyboard Support and Physical Keyboard Support. Also, the cursor works as expected for the web.

> One more point is that index.ios.js and index.js (IOUAmount) has a lot of duplicate code which can be dried up.

It certainly looks like that at first look. But the implementation of `updateAmount`, `updateAmountNumberPad` and `onSelectionChange` depend on the platform. Rest of the functions have been lifted up to `IOUAmountPage` to reduce redundancy. Let me know if there is still scope for reduction.
Thanks a lot for looking into it :D

sig5 commented:
I have confirmed by testing that using value in a controlled manner, resolves the bug as expected.
Can I add a prop `isValueControlled` that gives us control over if we want value to be passed down as prop or if we want to use `setNativeProps` to update value?

cc- @rushatgabhane

rushatgabhane commented:
> Can I add a prop isValueControlled in BaseTextInput that gives us control over if we want value to be passed down as prop or if we want to use setNativeProps to update value?

Sure, go ahead.

sig5 committed:
Added isValueControlled prop

sig5 committed:
Added isValueControlled Prop

sig5 commented:
> Yeah, correct. I may be confusing Android with Web. I was using the keyboard on the Android emulator while testing and I saw some strange behavior. But if you want, I can give it a test when you are done with the new changes.

I think I understand what behavior you are talking about. But that behavior is only there on the dev version. We have been testing android on the release version for the same reason :D.
> Just pointing out that this also fixes the issue. I know the set timeout rule but is a potential solution.

Thanks Rajat, let's try to fix it without setTimeout first, we can always pivot back to it :D

>https://github.com/sig5/App/blob/f50a882713fec3c5aef4d76fbcbfdf7f95695bb7/src/pages/iou/steps/IOUAmountPage/index.js#L98 If we are removing values then why do we need to validate the amount. And https://github.com/sig5/App/blob/f50a882713fec3c5aef4d76fbcbfdf7f95695bb7/src/pages/iou/steps/IOUAmountPage/index.js#L106

We actually might run into issues when there are 8 digits and two decimals. We don't want to update selection if user wants to delete the decimal :P

sig5 committed:
added defaultprop

rushatgabhane reviewed (dismissed):
@puneetlath ~~LGTM! Tests well on all platforms.~~

~~Note: Tests should be done on release variants of iOS and Android. Because on dev version, the cursor is misplaced on typing fast. This is due to the overhead of dev environment.~~

Update: not yet ready.

parasharrajat commented on src/components/TextInput/baseTextInputPropTypes.js:
```diff
@@ -14,6 +14,9 @@ const propTypes = {
     /** Default value - used for non controlled inputs */
     defaultValue: PropTypes.string,
 
+    /** False when Value set by setNativeProps */
```
This does not tell why this prop should be used at all. I think you should add a proper explanation and a warning to not use it unless needed. Give a reason why we have to introduce this prop. It is fine if the comment extends multiple lines.

rushatgabhane replied:
Thanks for taking a look at it Rajat

parasharrajat reviewed (changes requested):
please dry your changes. A lot of duplicate code is used. Just move the necessary code to a separate component that is different across platforms.  

A better name would be IOUAmountInput for the new component

sig5 committed:
Modified comment

sig5 committed:
Modified comment

sig5 committed:
Modified comment

sig5 committed:
Refactoring code

sig5 committed:
Added utility class

sig5 committed:
handled propTypes

sig5 committed:
fixed proptypes

rushatgabhane commented:
Thanks @sig5 and @parasharrajat, the code looks so much better now (Duplicates gone).
I'll give it a final review in a while

parasharrajat reviewed (changes requested):
Sorry, but refactor is going in a different direction. Utils should not be created for one specific page or component.

I will give you a few steps. Let me know if there are some challanges with it.
1. First, we will keep only one file for the page. IOUAmountPage. 
2. All the logic that is common will reside in this component. 
3. Then you create two files for IOUAmountInput (.ios.js and .js) which just manages the selection logic that we need to work out iOS. 

Don't forget to move the IOUAmountInput folder to IOUAmountPage folder. To do this you have to move the component to `IOUAmountPage/index.js` file.

sig5 commented:
Hi @parasharrajat ,
I have been looking to reduce redundancy for a while now. There might be some issues with the suggested approach. The selection logic if placed in `IOUAmountInput` will rather increase the redundancy as it will lead to duplication of logic in the render function and component update functions. (The changes you suggested are similar to the logic in the last set of changes). 

After thinking through the changes involved, this feels like the structure with the best readability and least redundancy. I agree about the moving the files part of the suggestion though. Can we maybe keep the `IOUAmountUtils` file while keeping it in the same folder?

rushatgabhane commented:
@parasharrajat Ofc there are multiple ways to DRY the code. But I think that with your suggestions, solving duplication of render function might lead to less readable code. I think we have dried up the code now, and it is also very easy to understand.

I agree with folder suggestions, and util is there but that's a fair compromise. Excited for your thoughts on this after you've played around with this. Thanks

sig5 committed:
Moved files

rushatgabhane commented on src/pages/iou/steps/IOUAmountPage/IOUAmountInput.js:
```diff
@@ -0,0 +1,170 @@
+ * iOS has incorrect selection(caret) position when using setNativeProps
+ * to update text and selection concurrently.
+ * Issue: https://github.com/facebook/react-native/issues/33077
+ */
```
@sig5 hmm, it should be sufficient to have this comment in  the iOS file only.

rushatgabhane commented on src/pages/iou/steps/IOUAmountPage/index.ios.js:
```diff
@@ -0,0 +1,84 @@
+        });
+    }
+
+
```
remove extra line here

rushatgabhane replied:
remove extra new line

rushatgabhane commented on src/pages/iou/steps/IOUAmountPage/index.js:
```diff
@@ -0,0 +1,81 @@
+        });
+    }
+
+
```
remove extra line here

rushatgabhane replied:
remove extra new line

rushatgabhane commented on src/pages/iou/steps/IOUAmountPage/IOUAmountPagePropTypes.js:
```diff
@@ -0,0 +1,15 @@
+
+    /** Previously selected amount to show if the user comes back to this screen */
+    selectedAmount: PropTypes.string.isRequired,
+
```
remove new line

rushatgabhane commented on src/pages/iou/steps/IOUAmountPage/index.ios.js:
```diff
@@ -0,0 +1,84 @@
+     * Validate new amount with decimal number regex up to 6 digits and 2 decimal digit to enable Next button
+     *
+     * @param {String} key
+     * @returns {String}
```
This function doesn't return a String, please update the docs.

rushatgabhane commented on src/pages/iou/steps/IOUAmountPage/IOUAmountUtils.js:
```diff
@@ -0,0 +1,114 @@
+ * @param {String} key -Key pressed
+ * @returns {Object}
+ */
+
```
remove extra new line

rushatgabhane commented on src/pages/iou/steps/IOUAmountPage/IOUAmountUtils.js:
```diff
@@ -0,0 +1,114 @@
+    return amount.replace(/,/g, '');
+}
+
+
```
remove extra new line

rushatgabhane commented on src/pages/iou/steps/IOUAmountPage/IOUAmountPagePropTypes.js:
```diff
@@ -0,0 +1,15 @@
+import PropTypes from 'prop-types';
+
+const propTypes = {
+
```
remove extra new line

rushatgabhane reviewed (changes requested):
@sig5 Dropping really small style changes.
* ~~Could you please move `IOUAmountInput` and it's related files to a folder?~~
* ~~Rename IOUAmountUtil -> IOUAmount*Input*Util.~~

Cheers!


Edit: got confused with the current structure first glance.

sig5 committed:
Resolved spaces comments and newlines

rushatgabhane commented:
@sig5 please merge `main` from upstream into your branch, and re-test all platforms because there have been many changes to `TextInput`. Let's also add screenshots for Safari

rushatgabhane reviewed (changes requested):
@sig5 please merge `main` from upstream into your branch, and re-test all platforms because there have been many changes to `TextInput`. Let's also add screenshots for Safari

rushatgabhane commented:
@sig5 how's it looking?

chiragsalian commented:
Hey there @sig5 👋 , so any update here? 
Just wanted to let you know that as per our [contributing guidelines](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md), we mention,
> Any issue that doesn't receive an update for 1 full week may be considered abandoned and the original contract terminated.

We'll give you this week to provide an update but if there is no update we'll have to consider the PR as abandoned and we won't be able to pay you for your work even if the remaining work is minimal. So anyway, if possible we're hoping you would still continue the task here since the only thing pending at this time is this [comment](https://github.com/Expensify/App/pull/7433#pullrequestreview-906546073) for you to address. You've done good work here so it would definitely be a shame if this was considered abandoned.

sig5 commented:
Hi @chiragsalian , 
 this PR has been particularly difficult to manage because of the `textInput` refactoring PR's which are being continuously merged over the past few weeks. The changes being done over those PR's remain a tad bit unclear due to lack of detailed plan/comment. These changes lead to breaking changes in the current PR. Every eventual merge leads to new issues while resolving this issue, and leaving the PR owner to analyse the whole set of changes right from the beginning.
 I appreciate and totally support the refactors taking place, but since its a bit too coupled with the PR, I was having issues with the PR. 
I would still like to work on this PR. I will post an update about potential changes that I might need to do to the `TextInput` over the couple next days, if that works fine?

sig5 commented:
That sounds good to me!

sig5 committed:
merged main

sig5 committed:
fixed ref

sig5 committed:
fixed naming

sig5 commented:
I have made a few changes that were an issue in the previous PR and merged main. The current PR tests well for me. The changes are merely modifications to the previous PR as per the modifications in main.

rushatgabhane reviewed (changes requested):
@sig5 bug: copy pasting amount doesn't work on iOS - Safari.

https://user-images.githubusercontent.com/29673073/160229403-581b5dc9-0b91-4d92-b091-e6de9071960c.mov

sig5 committed:
fixed safari copy paste issue

sig5 commented:
Hi @rushatgabhane, iOS mWeb seems to natively append a space before the pasted text. I have added a replace call using regex for tackling the issue.

rushatgabhane commented:
Thanks @sig5, could please add the copy paste thing in QA steps.

> Block updates
Single Updates
Invalid Updates (Length/Decimal etc)

And also please overexplain the QA steps so that someone unfamiliar with the issue can also test the PR, for all cases.

sig5 commented:
Done! @rushatgabhane

rushatgabhane commented:
@sig5 kindly fill in the PR checklist (it didn't exist when your PR was created)
https://github.com/Expensify/App/blob/main/.github/PULL_REQUEST_TEMPLATE.md

sig5 commented:
Done! @rushatgabhane

rushatgabhane reviewed (changes requested):
@sig5 the cursor position is wrong when copy pasting in Android.

rushatgabhane commented:
@sig5 any updates?

sig5 commented:
Hi, I have been looking at this issue for a while. So this caret positioning seems to be off on android and iOS- Safari mWeb. But since the same code works well on all platforms that use `index.js` .  This seems to me like an native issue. 
Imho the bug appears in rare use-cases (On pasting, cursor doesn't go to end of selection on iOS and it goes to end of text on iOS-Safari) . I have been trying to find the cause but am unable to find the exact cause, so please let me know the next steps.

rushatgabhane commented:
@sig5 thanks for your patience, I'll get back to this by tomorrow EOD.

rushatgabhane commented:
> I have been trying to find the cause but am unable to find the exact cause, so please let me know the next steps.

@sig5 cursor position after pasting in compose box is correct on all platforms. So can you please help us understand why it doesn't work for IOUAmount page?

rushatgabhane commented:
@sig5 / @chiragsalian this PR can be closed as it's no longer being worked on.
